presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,200 @@ public void close()
+                        unfinishedRecoverableLifespanStageStates.remove(lifespanStage);
+                        partitionCommitter.commitPartition(lifespanStage.getLifespan().getId(), lifespanStageState.getFragments());
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unrecognized commit granularity: " + tableCommitContext.getCommitGranularity());
+            }
+        }
+
+        public List<Page> getStatisticsPagesToProcess(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanStage lifespanStage = LifespanStage.fromTableCommitContext(tableCommitContext);
+            switch (tableCommitContext.getCommitGranularity()) {
+                case TABLE:
+                    return extractStatisticsRows(page).map(ImmutableList::of).orElse(ImmutableList.of());
+                case PARTITION:
+                    if (!finishedRecoverableLifespanStages.containsKey(lifespanStage)) {
+                        return ImmutableList.of();
+                    }
+                    return finishedRecoverableLifespanStages.get(lifespanStage).getStatisticsPages();
+                default:
+                    throw new IllegalStateException("Unrecognized commit granularity: " + tableCommitContext.getCommitGranularity());
+            }
+        }
+
+        public long getRowCount()
+        {
+            return Stream.concat(unrecoverableLifespanStageStates.values().stream(), finishedRecoverableLifespanStages.values().stream())
+                    .mapToLong(LifespanStageState::getRowCount)
+                    .sum();
+        }
+
+        public List<Slice> getFragments()
+        {
+            return Stream.concat(unrecoverableLifespanStageStates.values().stream(), finishedRecoverableLifespanStages.values().stream())
+                    .map(LifespanStageState::getFragments)
+                    .flatMap(List::stream)
+                    .collect(toImmutableList());
+        }
+
+        private TableCommitContext getTableCommitContext(Page page)
+        {
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                Block operatorExecutionContextBlock = page.getBlock(CONTEXT_CHANNEL);
+                if (!operatorExecutionContextBlock.isNull(position)) {
+                    return tableCommitContextCodec.fromJson(operatorExecutionContextBlock.getSlice(position, 0, operatorExecutionContextBlock.getSliceLength(position)).getBytes());
+                }
+            }
+            throw new IllegalStateException("No operator execution context in input page");
+        }
+
+        private static boolean isLastPageForGroupedLifespan(Page page)
+        {
+            return !page.getBlock(ROW_COUNT_CHANNEL).isNull(0);
+        }
+
+        private static class LifespanStage
```

Comment:
nit: just call it `LifespanAndStage`

Reply:
I'm considering `StagedLifespan`...How is that?